weave/integrations/dspy/dspy_utils.py

```diff
@@ -0,0 +1,129 @@
+import importlib
+from typing import TYPE_CHECKING, Any, Callable, Union
+
+import weave
+from weave.integrations.patcher import SymbolPatcher
+from weave.trace.autopatch import OpSettings
+from weave.trace.op import Op
+from weave.trace.serialization.serialize import dictify
+
+if TYPE_CHECKING:
+    from dspy.primitives.prediction import Example
+
+
+def get_symbol_patcher(
+    base_symbol: str, attribute_name: str, settings: OpSettings
+) -> SymbolPatcher:
+    display_name = base_symbol + "." + attribute_name
+    display_name = (
+        display_name.replace(".__call__", "")
+        if attribute_name.endswith(".__call__")
+        else display_name
+    )
+    return SymbolPatcher(
+        lambda: importlib.import_module(base_symbol),
+        attribute_name,
+        dspy_wrapper(
+            settings.model_copy(update={"name": settings.name or display_name})
+        ),
+    )
+
+
+def serialize_dspy_objects(data: Any) -> Any:
+    import numpy as np
+    from dspy import Example, Module
+
+    if isinstance(data, Example):
+        return data.toDict()
+
+    elif isinstance(data, Module):
+        return data.dump_state()
+
+    elif isinstance(data, np.ndarray):
+        return data.tolist()
+
+    elif isinstance(data, dict):
+        return {key: serialize_dspy_objects(value) for key, value in data.items()}
+
+    elif isinstance(data, list):
+        return [serialize_dspy_objects(item) for item in data]
+
+    return data
+
+
+def dspy_postprocess_inputs(inputs: dict[str, Any]) -> dict[str, Any]:
+    from dspy import Adapter, Evaluate, Predict
+
+    if "self" in inputs:
+        dictified_inputs_self = dictify(inputs["self"])
+
+        # Serialize the signature of the object if it is a Predict or Adapter
+        if isinstance(inputs["self"], Predict) or isinstance(inputs["self"], Adapter):
+            if hasattr(inputs["self"], "signature"):
+                if hasattr(inputs["self"].signature, "model_json_schema"):
+                    dictified_inputs_self["signature"] = inputs[
+                        "self"
+                    ].signature.model_json_schema()
+                else:
+                    dictified_inputs_self["signature"] = inputs["self"].signature
```

Comment:
I assume this is a non-pydantic object?  Not sure what signature means for dspy... I would have thought it was the `inspect.signature`

Reply:
Yes, this is a DSPy-specific paradigm ([ref](https://dspy.ai/learn/programming/signatures/)), not `inspect.signature`.